include/seastar/rpc/rpc.hh

```diff
@@ -112,6 +112,21 @@ struct client_options {
     sstring isolation_cookie;
 };
 
+struct send_options {
+    /// Specify whether this RPC call is interested in the RPC module reporting that there was a connection error.
+    /// Seastar will log a connection error if and only if either:
+    /// - there was at least one one-way RPC sent on this connection which specified report_connection_error=true,
+    /// - or there was at least one two-way RPC sent on this connection which specified report_connection_error=true
+    ///   and didn't return yet.
+    ///
+    /// Note: this works only for sending RPCs, not receiving them. That is, if there is a connection on which we didn't
+    /// send any RPCs with report_connection_error=true, but the other side did (and we may have sent responses),
+    /// and an error happens, we won't report it.
+    ///
+    /// Note: streams cannot be configured in this way and always report connection errors.
+    bool report_connection_error = true;
+};
+
```

Comment:
OTOH this would complicate the caller side. Consider Scylla messaging_service. Right now we create a client and cache it immediately, and other RPCs can immediately start using the client - even before it connected underneath. If it returned a future<> and we waited on it, we would need to deal with those other RPCs somehow, e.g. by queuing them up and flushing the queue when the future<> becomes available.

Reply:
Also, the patch silences errors for clients that did manage to connect (but executed only the subset of RPCs which are not interested in the log).

Reply:
I don't see any reason for rpc to log anything. It should return any errors to its callers via an exception, who can then decide whether to log or not. We have to log now because the constructor has no way to return asynchronous errors.

wrt. Scylla, clients_map would map into a variant<in-progress connection, established connection>. Yes, it moves the complexity from here to Scylla, but the fact that we are moving decisions about logging from Scylla to here with elaborate control shows that swallowing the errors here is not a good idea.

Reply:
> I don't see any reason for rpc to log anything. It should return any errors to its callers via an exception, who can then decide whether to log or not. We have to log now because the constructor has no way to return asynchronous errors.
> 
There may be million simultaneous callers. Do you want each one of them to report a failure? Even if you do it will be hard to change it now since all of them ignore rpc::closed_errors because it is logged once for entire connection and not per rpc call.

> wrt. Scylla, clients_map would map into a variant<in-progress connection, established connection>. Yes, it moves the complexity from here to Scylla, but the fact that we are moving decisions about logging from Scylla to here with elaborate control shows that swallowing the errors here is not a good idea.

I is trivial to log errors only after the connection is established. There is another PR that does exactly that. This one tried to do something more complicated and unnecessary.

Reply:
> > I don't see any reason for rpc to log anything. It should return any errors to its callers via an exception, who can then decide whether to log or not. We have to log now because the constructor has no way to return asynchronous errors.
> 
> There may be million simultaneous callers. Do you want each one of them to report a failure? Even if you do it will be hard to change it now since all of them ignore rpc::closed_errors because it is logged once for entire connection and not per rpc call.
> 

It's up to the callers to decide what to log. An evolved application may decide to log nothing.

btw, we should return std::expected for network errors, exceptions are too expensive for such easily triggered conditions.

> > wrt. Scylla, clients_map would map into a variant<in-progress connection, established connection>. Yes, it moves the complexity from here to Scylla, but the fact that we are moving decisions about logging from Scylla to here with elaborate control shows that swallowing the errors here is not a good idea.
> 
> I is trivial to log errors only after the connection is established. There is another PR that does exactly that. This one tried to do something more complicated and unnecessary.

How does RPC know which errors are okay to log an which are not?